### PR TITLE
Same binaries for manoto and non-manoto and upload signed executables closes #3007

### LINKED ISSUE #3007: Manoto build process should not rebuild binaries themselves and should upload signed binaries instead of unsigned
The current Manoto build process also rebuilds the binaries themselves. This is problematic because those binaries are slightly different and don't have the same exact data. Instead, the Manoto build should purely take existing binaries and should repackage them.

### PR DESCRIPTION
**Pull request by myleshorton**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -331,11 +333,11 @@ package-windows: require-version windows
 	if [[ -z "$$SECRETS_DIR" ]]; then echo "SECRETS_DIR environment value is required."; exit 1; fi && \
 	if [[ -z "$$BNS_CERT_PASS" ]]; then echo "BNS_CERT_PASS environment value is required."; exit 1; fi && \
 	$(call docker-up) && \
-	docker run -v $$PWD:/lantern -v $$SECRETS_DIR:/secrets -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && BNS_CERT="/secrets/bns.pfx" BNS_CERT_PASS="'$$BNS_CERT_PASS'" VERSION="'$$VERSION'" MANOTO="'$$MANOTO'" make docker-package-windows' && \
-	echo "-> lantern-installer.exe"
+	docker run -v $$PWD:/lantern -v $$SECRETS_DIR:/secrets -t $(DOCKER_IMAGE_TAG) /bin/bash -c 'cd /lantern && BNS_CERT="/secrets/bns.pfx" BNS_CERT_PASS="'$$BNS_CERT_PASS'" VERSION="'$$VERSION'" MANOTO="'true'" make docker-package-windows' && \
```
Comment (xiam) [inline Makefile]: Where did this `bns.pfx` file go?

Comment (myleshorton) [inline Makefile]: It's actually not checked in anywhere -- it's a newly signed file that I'm keeping out of even private repos for now.

File: src/github.com/getlantern/yamlconf/yamlconf_test.go [hunk 2]
```diff
@@ -214,69 +213,6 @@ func TestCustomPoll(t *testing.T) {
...
-
-	assert.Equal(t, &TestCfg{
-		Version: 3,
-		N: &Nested{
-			S: newNested.S,
-			I: FIXED_I,
-		},
-	}, updated, "Nested I should have reverted to default value after clearing")
-}
-
```
Comment (myleshorton) [inline src/github.com/getlantern/yamlconf/yamlconf_test.go]: This was testing old functionality that's now stripped. This should have been removed in that ticket but somehow slipped through.

**Comment by xiam** [comment]
I tested with a random pfx file, it looks like the `lantern_darwin_amd64` is different from the lantern `binary`:

```
> gsha1sum /Applications/Lantern.app/Contents/MacOS/lantern
656d31e3638d8af9e3ed00c7ab9425bf8c2f7162  /Applications/Lantern.app/Contents/MacOS/lantern

> gsha1sum Lantern.app/Contents/MacOS/lantern
656d31e3638d8af9e3ed00c7ab9425bf8c2f7162  Lantern.app/Contents/MacOS/lantern

> gsha1sum lantern_darwin_amd64
8e7a1710dc8f9d1707f26f8684c580131c85efa7  lantern_darwin_amd64
```

**Comment by xiam** [comment]
The same happens with Windows:

```
gsha1sum update_windows_386
2eb8b2896a50455061c4e90554ec67d051503b6c  update_windows_386

gsha1sum lantern.exe
f3d6dbb4d8caa2a17ddd2b4048750f6aed9a7673  /Users/rev/Shared/lantern.exe
```

But it does not happen on Linux.

Does signing the installer change the binary?

**Comment by myleshorton** [comment]
Yeah signing the installer definitely changes the binary on OSX and
windows, which is what this is designed to fix. Sounds like there's a bug
though!

**Comment by xiam** [comment]
Windows has 3 different hashes, in this case:

`lantern_windows_386.exe` is unsigned.
`update_windows_386` is signed.
`lantern.exe` was affected by the installer signature.

Is that right?

**Comment by myleshorton** [comment]
Oh interesting - I was unaware of #3 but that could explain what you're
seeing.

**Comment by myleshorton** [comment]
So just to be clear on all the binary differences @xiam, the new changes don't change lantern_windows_386.exe, update_windows_386, or lantern_darwin_amd64 but just the bzip2 files (i.e. the bzip2 files should contain the signed versions).

**Comment by myleshorton** [comment]
So you should be comparing the binaries in the compressed files only.

**Comment by xiam** [comment]
Yeah, I uncompressed the `update_windows_386.bz2` file:

```
gsha1sum update_windows_386
2eb8b2896a50455061c4e90554ec67d051503b6c  update_windows_386

gsha1sum lantern.exe
f3d6dbb4d8caa2a17ddd2b4048750f6aed9a7673  /Users/rev/Shared/lantern.exe
```

`lantern.exe` is the binary that the installer creates, I think that is going to be problematic because those files have two different hashes, the autoupdate will take `update_windows_386` and will replace the old binary with it, leaving a file with a `2eb8b2896a50455061c4e90554ec67d051503b6c` checksum instead of the original `f3d6dbb4d8caa2a17ddd2b4048750f6aed9a7673`. I am wondering if we could extract the `lantern.exe` somehow after signing the installer, that's what I've been doing manually for our releases.

**Comment by myleshorton** [comment]
Thanks @xiam. I just realize I get this in docker-package-windows

```
/bin/bash: line 5: bzip2: command not found
```

I think I have to add that to the Dockerfile huh?

**Comment by xiam** [comment]
Yes @myleshorton

**Comment by myleshorton** [comment]
Cool just added it.

**Comment by myleshorton** [comment]
I actually get the same value for each:

On OSX:

```
shasum update_windows_386
0a6f68d238e38ae3baf03985a14397322ff588ac  update_windows_386
```

On Win:

```
Adam Fisk@WIN-2M0SC8OPGJ6 ~
$ shasum /cygdrive/c/Users/Adam\ Fisk/AppData/Roaming/Lantern/lantern.exe
0a6f68d238e38ae3baf03985a14397322ff588ac */cygdrive/c/Users/Adam Fisk/AppData/Roaming/Lantern/lantern.exe
```

**Comment by xiam** [comment]
Hmm, that's weird. I'm going to `make clean` and check again.

**Comment by xiam** [comment]
Same thing unfortunately. Does this look right?

```
MANOTO=true BNS_CERT_PASS=foofoo SECRETS_DIR=~/projects/lantern/too-many-secrets VERSION=2.0.0 make packages
...
Generating distribution package for darwin/amd64 manoto...
Packaged settings: startupurl: https://www.facebook.com/manototv/app_128953167177144
/bin/bash: Lantern.app/Contents/Resources/en.lproj/.packaged-lantern.yaml: No such file or directory
make: *** [package-darwin-manoto] Error 1
```

```
git branch | grep "^*"
* issue3007

git log -n1
commit d3f2815399c2d3cdb13d788b6c7fd13b7b967603
Author: Myles Horton <myles@getlantern.org>
Date:   Fri Aug 28 16:14:43 2015 -0700

    added bzip


find . | grep packaged-lan
./32bit/usr/lib/lantern/.packaged-lantern.yaml
./64bit/usr/lib/lantern/.packaged-lantern.yaml
./installer-resources/windows/.packaged-lantern.yaml
```

**Comment by myleshorton** [comment]
Do you have this directory?

installer-resources/darwin/Lantern.app_template/Contents/Resources/en.lproj/

**Comment by xiam** [comment]
I initially did not have it:

```
ls Lantern.app/Contents/Resources
app.icns
```

So I created it and ran the make again:

```
mkdir Lantern.app/Contents/Resources/en.lproj
make ...
```

Failed again at the same point:

```
...
/bin/bash: Lantern.app/Contents/Resources/en.lproj/.packaged-lantern.yaml: No such file or directory
make: *** [package-darwin-manoto] Error 1
``` 

And the directory en.lproj was gone again.

```
ls Lantern.app/Contents/Resources/
app.icns
```

Is my make invocation correct? I also tried `MANOTO=1`, same results.

I am using a self signed `.pfx` file, maybe that affects something? I'm going to give it a closer look.

**Comment by myleshorton** [comment]
It has to be in Lantern.app_template

**Comment by myleshorton** [comment]
The pfx doesn't affect OSX at all. I think it's the en.lproj not being there in the template.

**Comment by myleshorton** [comment]
I think that last change should do it @xiam

**Comment by xiam** [comment]
Great! the latest commit worked and the checksum issue is not there anymore:

```
open lantern-installer.dmg
gsha1sum /Applications/Lantern.app/Contents/MacOS/lantern
f21e4e682b43ebff74e20df6c6c7ac1a3e76f22d  /Applications/Lantern.app/Contents/MacOS/lantern
bzip2 -d update_darwin_amd64.bz2
gsha1sum update_darwin_amd64
f21e4e682b43ebff74e20df6c6c7ac1a3e76f22d  update_darwin_amd64
```

```
gsha1sum lantern.exe
ac04b0856a6330edb0d7e9e20671ec8080a2ef7b  lantern.exe
bzip2 -d update_windows_386.bz2
gsha1sum update_windows_386
ac04b0856a6330edb0d7e9e20671ec8080a2ef7b  update_windows_386
```

Looks good to me, I only added a yum clean all at the end of the bzip2 installation. See https://github.com/getlantern/lantern/pull/3067

**Comment by myleshorton** [comment]
Awesome! And windows is working for you too?

**Comment by xiam** [comment]
Yep, Windows matches as well! The thing that is not working now is Linux, bot 64 and 32bit. Could you test this to see if this works for you? You can install "dpkg" with brew.

```
dpkg -x lantern_2.0.1_amd64.deb 64bit
bzip2 -d update_linux_amd64.bz2
gsha1sum update_linux_amd64
e6115fb974057b2c1ababd8f3204c862836b1e22  update_linux_amd64
gsha1sum 64bit/usr/lib/lantern/lantern-binary
507ea671f78bacd32e7efcd17b78f949e894c70d  64bit/usr/lib/lantern/lantern-binary
dpkg -x lantern-manoto_2.0.1_amd64.deb 64bit-manoto
gsha1sum 64bit-manoto/usr/lib/lantern/lantern-binary
507ea671f78bacd32e7efcd17b78f949e894c70d  64bit-manoto/usr/lib/lantern/lantern-binary
gsha1sum lantern_linux_amd64
507ea671f78bacd32e7efcd17b78f949e894c70d  lantern_linux_amd64
```

Why does the uncompressed `update_linux_amd64.bz2` compute a different signature?

**Pull request merged**